### PR TITLE
Refactor `NotebookCellOutline` layers to remove `target` + breadcrumb fix

### LINKED ISSUE #211812: Notebook breadcrumbs don't respect show code cell setting
Type: <b>Bug</b>

When using notebook breadcrumbs, the setting `notebook.breadcrumbs.showCodeCells` is not respected. Code cells will always be present in the breadcrumbs.

I believe this is caused along the way with the outline perf work, since there is now only a single outline provider with a target `OutlinePane` which filters out code cells at a higher layer independently.

VS Code version: Code - Insiders 1.89.0-insider (b58957e67ee1e712cebf466b995adf4c5307b2bd, 2024-05-01T02:05:10.904Z)
OS version: Windows_NT x64 10.0.22621
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|12th Gen Intel(R) Core(TM) i7-1265U (12 x 2688)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|undefined|
|Memory (System)|31.83GB (7.18GB free)|
|Process Argv|--crash-reporter-id 1cb7488a-bb2c-46a8-ac2f-3151dbb5b996|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (21)</summary>

Extension|Author (truncated)|Version
---|---|---
better-comments|aar|3.0.2
github-markdown-preview|bie|0.3.0
vscode-eslint|dba|2.4.4
gitlens|eam|2024.5.105
prettier-vscode|esb|10.4.0
copilot|Git|1.184.849
copilot-chat|Git|0.15.2024043005
vscode-pull-request-github|Git|0.87.2024042613
debugpy|ms-|2024.5.11172029
python|ms-|2024.5.11172159
vscode-pylance|ms-|2024.4.105
jupyter|ms-|2024.4.2024042601
jupyter-keymap|ms-|1.1.2
jupyter-renderers|ms-|1.0.17
vscode-jupyter-cell-tags|ms-|0.1.9
vscode-jupyter-slideshow|ms-|0.1.6
remote-containers|ms-|0.364.0
remote-wsl|ms-|0.88.2
vscode-github-issue-notebooks|ms-|0.0.130
ayu|tea|1.0.5
vscodeintellicode|Vis|1.3.1

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vspor879:30202332
vspor708:30202333
vspor363:30204092
vscod805:30301674
vsaa593:30376534
py29gd2263:31024238
c4g48928:30535728
a9j8j154:30646983
962ge761:30841072
pythongtdpath:30726887
welcomedialog:30812478
pythonidxpt:30768918
pythonnoceb:30776497
asynctok:30898717
dsvsc013:30777762
dsvsc014:30777825
dsvsc015:30821418
pythontestfixt:30866404
pythonregdiag2:30926734
pyreplss1:30879911
pythonmypyd1:30859725
pythoncet0:30859736
2e7ec940:31000449
pythontbext0:30879054
accentitlementst:30870582
dsvsc016:30879898
dsvsc017:30880771
dsvsc018:30880772
cppperfnew:30980852
fegfb526:30952798
bg6jg535:30979844
ccp1r3:30958150
pythonait:30973460
gee8j676:31009450
showvideot:31016890
chatpanelt:31014475
bdiig495:31013172
pythoncenvpt:31022790
a69g1124:31018687
j0bi8933:31021562
pythonprc:31030355
dwnewjupytercf:31035177

```

</details>

### LINKED ISSUE #211998: Notebook breadcrumbs are tied to OutlinePane filters
This is related to `recomputeActive` inside the notebook outline provider. Right now, since this was previously acting as a filter layer, turning off code cells for the outline pane would prevent a recompute of the active element. 

🐛 if you turn off code cells for the outline pane, you will not get any recomputing of the breadcrumbs as you click around code cells. it will only be triggered if you click on an md cell.

🐛 the notebook setting gotosymbols show all symbols is only ever respected when code cells is turned on in the outline

### PR DESCRIPTION
**Pull request by Yoyokrazy**
Fixes: #211998, #211812

This is conveniently part of the ongoing outline refactor to pull filters out of the primary outline provider, in favor of filtering at higher layers. 

Previously, the `recomputeActive()` call within `NotebookCellOutlineProvider` was only resetting the outline `activeElement` according to passing the three `outlinePane` filters for headers, code cells, and symbols. This `activeElement` was also used by the `breadcrumbsDataSource`, which meant that if you have code cells filtered out of your outline pane, 🐛 you will never see your breadcrumbs update as long as you click through code cells (because the active element is never getting set).

The fix:
- remove filters from the provider. `NotebookCellOutlineProvider.activeElement` will now always hold the active entry, even if it will be filtered down the line.
- add in a new getter for `activeBreadcrumbsElement`. This allows us to filter code cells out of the outline pane while still showing them in breadcrumbs (if that's something someone would ever do... regardless.)

New behavior to the outline pane:
This is only really relevant if you have `Follow Cursor` enabled for the view. Previously, if you were clicking around in cells that weren't directly displayed in the outline, the highlight in the outline just wouldn't move. Now, the selection will shift to the nearest valid parent `OutlineEntry` of the cell you selected. Makes this feature a touch more useful.

Current behavior:

https://github.com/microsoft/vscode/assets/12552271/2284f43a-2576-43f9-8c67-11e8a2d1be84

New Proposed behavior:

https://github.com/microsoft/vscode/assets/12552271/220435ca-e2ce-4df3-b55e-dbb5264d9a44

### CODE REVIEW

**Review by kalebers — APPROVED** [review]

**Comment by Yoyokrazy** [comment]
Latest commit includes more refactor work:
- OutlineTarget is removed from all layers except the class `NotebookCellOutline`
  - target here only controls config and options for the outline trees, the outline renderer, and setFullSymbols (which will need to be moved anyway in a further refactor that addresses more perf work)
- the 3 data sources for the IOutline now just hold a reference to the primary `NotebookCellOutlineDataSource`
- bit of fixing for testing to reflect shape changes of the NotebookCellOutline and it's view providers/data sources

File: src/vs/workbench/contrib/notebook/browser/contrib/outline/notebookOutline.ts [hunk 1]
```diff
@@ -334,17 +334,64 @@ export class NotebookQuickPickProvider implements IQuickPickDataSource<OutlineEn
...
+		}
+
+		if (!this.filterEntry(newActive)) {
+			return newActive;
+		}
+
+		// find a valid parent
+		let parent = newActive.parent;
+		while (parent) {
+			if (this.filterEntry(parent)) {
```
Comment (rebornix) [inline src/vs/workbench/contrib/notebook/browser/contrib/outline/notebookOutline.ts:354]: Some potential perf optimization:

- we are reading 3 settings each time when we read `filterEntry`, they can probably be cached as they won't change in this while loop.
- minor: maybe we can cache the active entry so we don't need to recompute if `activeElement` doesn't change.

**Review by rebornix — CHANGES_REQUESTED** [review]

**Comment by Yoyokrazy** [comment]
Diagram of the new architecture for notebook outline (+ related views):
image

**Review by rebornix — APPROVED** [review]
Looks great!

**Pull request merged**